flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/input/MultiStateKeyIterator.java

```diff
@@ -31,65 +31,82 @@
+ * completely enumerating and buffering nested iterators event for a single call to {@link
+ * MultiStateKeyIterator#hasNext}.
+ *
+ * @see <a
+ *     href="https://bugs.openjdk.org/browse/JDK-8267359">https://bugs.openjdk.org/browse/JDK-8267359</a>
  * @param <K> Type of the key by which state is keyed.
  */
 @Internal
-public final class MultiStateKeyIterator<K> implements CloseableIterator<K> {
+public class MultiStateKeyIterator<K> implements CloseableIterator<K> {
     private final List<? extends StateDescriptor<?, ?>> descriptors;
 
     private final KeyedStateBackend<K> backend;
 
-    private final Iterator<K> internal;
+    private Iterator<? extends StateDescriptor<?, ?>> outerIter;
+    private Iterator<K> innerIter;
 
     private final CloseableRegistry registry;
 
     private K currentKey;
 
     public MultiStateKeyIterator(
             List<? extends StateDescriptor<?, ?>> descriptors, KeyedStateBackend<K> backend) {
+
+        outerIter = descriptors.iterator();
+        innerIter = null;
+
         this.descriptors = Preconditions.checkNotNull(descriptors);
         this.backend = Preconditions.checkNotNull(backend);
 
         this.registry = new CloseableRegistry();
-        this.internal =
-                descriptors.stream()
-                        .map(
-                                descriptor ->
-                                        backend.getKeys(
-                                                descriptor.getName(), VoidNamespace.INSTANCE))
-                        .peek(
-                                stream -> {
-                                    try {
-                                        registry.registerCloseable(stream::close);
-                                    } catch (IOException e) {
-                                        throw new RuntimeException(
-                                                "Failed to read keys from configured StateBackend",
-                                                e);
-                                    }
-                                })
-                        .flatMap(stream -> stream)
-                        .iterator();
+    }
+
+    @Override
+    public void close() throws Exception {
+        registry.close();
     }
 
     @Override
     public boolean hasNext() {
-        return internal.hasNext();
+        if ((innerIter == null || !innerIter.hasNext()) && outerIter.hasNext()) {
+            StateDescriptor<?, ?> descriptor = outerIter.next();
+            Stream<K> stream = backend.getKeys(descriptor.getName(), VoidNamespace.INSTANCE);
+            innerIter = stream.iterator();
+            try {
+                registry.registerCloseable(stream::close);
+            } catch (IOException e) {
+                throw new RuntimeException("Failed to read keys from configured StateBackend", e);
+            }
+        }
+        if (innerIter == null) {
+            return false;
+        }
+        return innerIter.hasNext();
```

Comment:
IIUC, this is incorrect.
After first if, innerIter maybe an empty iterator (due to lacking related states or empty states) but we could not say it doesn't hasNext (outerIter may still hasNext)

Reply:
You are right, thanks. In case of state sets the inner iterator always has at least one element otherwise it wouldn't be in the outer enumeration. How ever this assumption would be a tight coupling to external components. Should I rephrase the section ?

Reply:
Yeah, I think you should.
IIUC, just in this case, it may also bring some exceptions:
1. `StateDescriptor` is passed by users, which may not be correct
2. if user just registered state but not put any states, the state will be empty (only state meta is stored in the checkpoint, this also works)

Of course, even if it works till now, we also should guarantee that this class could work not coupling to other calsses.